scipy/stats/morestats.py

```diff
@@ -2078,6 +2078,54 @@ def anderson_ksamp(samples, midrank=True):
 AnsariResult = namedtuple('AnsariResult', ('statistic', 'pvalue'))
 
 
+
+class _ABW:
+    '''Distribution of Ansari-Bradley W-statistic under the null hypothesis'''
+    # TODO: calculate exact distribution considering ties
+    # We could avoid summing over more than half the frequencies,
+    # but inititally it doesn't seem worth the extra complexity
+
+    def __init__(self):
+        '''Minimal initializer'''
+        self.m = None
+        self.n = None
+
+    def pmf(self, k, m, n):
+        '''Probability mass function'''
+        self._recalc(m, n)
+        # convention is that PMF at k = 12.5 is the same as at k = 12
+        ind = int(k - self.astart)
+        return self.freqs[ind] / self.total
+
+    def cdf(self, k, m, n):
+        '''Cumulative distribution function'''
+        self._recalc(m, n)
+        # this is a bit subtle; easiest to think through example:
+        # the CDF at k = 12.5 should be the same as at k = 12
+        ind = int(np.floor(k - self.astart))
+        return self.freqs[:ind+1].sum() / self.total
+
+    def sf(self, k, m, n):
+        '''Survival function'''
+        self._recalc(m, n)
+        # this is a bit subtle; easiest to think through example:
+        # the SF at k = 12.5 should be the same at k = 13
+        ind = int(np.ceil(k - self.astart))
+        return self.freqs[ind:].sum() / self.total
+
+    def _recalc(self, m, n):
+        '''If necessary, recalculate exact distribution'''
+        if m != self.m or n != self.n:
+            astart, a1, _ = statlib.gscale(n, m)
+            self.astart = astart
+            self.freqs = a1.astype(np.float64)
+            self.total = self.freqs.sum()  # could calculate from m and n
+
+
+# Maintain state for faster repeat calls to ansari w/ method='exact'
+_abw_state = _ABW()
```

Comment:
This may be overkill for now, but it will be useful if we implement the distribution with ties.

Reply:
I agree that this can't hurt + this is better as, unlike before, it doesn't recalculate the distribution for repeated calls (with the same `n` and `m`) to `ansari`!